src/Microsoft.Graph.Core/Requests/RetryHandler.cs

```diff
@@ -0,0 +1,193 @@
+            while (retryCount < MAX_RETRY)
+            {
+
+                // Call Delay method to get delay time from response's Retry-After header or from exponential backoff 
+                // Start Task.Delay task
+                Task delay = Delay(response, retryCount);
+
+                // Get the original request
+                var request = response.RequestMessage;
+
+                // Increase retryCount and then update Retry-Attempt in request header
+                retryCount++;
+                AddOrUpdateRetryAttempt(request, retryCount);
+
+                // Delay time
+                await delay;
+
+                // Call base.SendAsync to send the request
+                response = await base.SendAsync(request, cancellationToken);
+
+                if (!IsRetry(response) || !IsBuffed(request))
+                {
+                    return response;
+                }
+
+            }
+            throw new ServiceException(
+                         new Error
+                         {
+                             Code = ErrorConstants.Codes.TooManyRetries,
+                             Message = string.Format(ErrorConstants.Messages.TooManyRedirectsFormatString, retryCount)
+                         });
+
+        }
+
+
+
+        /// <summary>
+        /// Check the HTTP response's status to determine whether it should be retried or not.
+        /// </summary>
+        /// <param name="response">The <see cref="HttpResponseMessage"/>returned.</param>
+        /// <returns></returns>
+        public bool IsRetry(HttpResponseMessage response)
+        {
+            if ((response.StatusCode == HttpStatusCode.ServiceUnavailable ||
+                response.StatusCode == (HttpStatusCode)429))
+            {
+                return true;
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// Check the HTTP request's content to determine whether it can be retried or not.
+        /// </summary>
+        /// <param name="request">The <see cref="HttpRequestMessage"/>needs to be sent.</param>
+        /// <returns></returns>
+        private bool IsBuffed(HttpRequestMessage request)
+        {
+            HttpContent content = request.Content;
+
+            if ((request.Method == HttpMethod.Put || request.Method == HttpMethod.Post) 
+                && (content == null || content.Headers.ContentLength == null || (int)content.Headers.ContentLength == -1))
+            {
+                return false;
+            }
+            return true;
+           
+        }
+
+        /// <summary>
+        /// Update Retry-Attempt header in the HTTP request
+        /// </summary>
+        /// <param name="request">The <see cref="HttpRequestMessage"/>needs to be sent.</param>
+        /// <param name="retry_count">Retry times</param>
+        private void AddOrUpdateRetryAttempt(HttpRequestMessage request, int retry_count)
+        {
+            if (request.Headers.Contains(RETRY_ATTEMPT))
```

Comment:
If the header is already there, you should be able to just update it rather than removing it and re-adding it